✨ create いっぱい追加してしまった…

shiki-01 opened the pull request:
・プレビュー
・Icon使えるように
・リストきれいに
・ハイライトコードつけれるように

色々追加してしまったので、一応こっちも確認もらおうかと…！

see:

- [`src/lib/utils/blogParsre`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/src/lib/utils/services/blogParser.ts)　ブログを一旦パーサーかけてからそれぞれ描写してます
- [`src/routes/blogs/[slug]/+page.server.ts`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/src/routes/blogs/%5Bslug%5D/%2Bpage.server.ts)　画面プレビュー用に draftKey があれば [`getDraftContentDetail`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/src/lib/utils/services/microcms.ts#L52)を実行するように変更しました
- [`src/routes/blogs/[slug]/+page.svelte`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/src/routes/blogs/%5Bslug%5D/%2Bpage.svelte)　Icon とか ハイライト とか色々デザイン面の変更です

メインはここらへんです！

shiki-01 committed:
✨ create いっぱい追加してしまった…

・プレビュー
・Icon使えるように
・リストきれいに
・ハイライトコードつけれるように

shiki-01 committed:
♻️ update リファクタリング

shiki-01 committed:
📁 save こっそり追加

microCMSに入ってる装飾系を大体追加

shiki-01 committed:
♻️ update もっと細かく

まだ途中 -> table

shiki-01 commented:
思ったより問題が見つかってしまったのでもうしばしお待ちして欲しいです！

終わりしだい連絡します🙇‍♂️

shiki-01 committed:
save

むり！

shiki-01 committed:
🐛 fix テーブルできた！

shiki-01 committed:
🐛 fix imageの修正

shiki-01 committed:
♻️ update tableをちょっとしゅうせい

shiki-01 committed:
✨ create いっぱいコンポーネント化

shiki-01 committed:
💠 スタイルの微調整＆完成！

shiki-01 committed:
そういえばもう使ってなかった

shiki-01 committed:
そういえば放置してた

shiki-01 committed:
そういえばelseだった

shiki-01 committed:
🐛 fix preの修正

shiki-01 committed:
📥 add tipsの追加＋CodeBlockの追加

shiki-01 commented:
色々制限したりオリジナルクラスも追加したので、microCMS 関連の tips を追加しました！
一応こちらも確認お願いします。

see:

- [`docs/microcms/tips.md`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/docs/microcms/tips.md)

shiki-01 committed:
♻️ update スペースが足りなかった

shiki-01 committed:
♻️ update リンクミスってた

shiki-01 committed:
♻️ update リンクの表記修正

shiki-01 committed:
♻️ update もうちょっと修正

shiki-01 committed:
♻️ update テーブルあたりのスタイル修正

shiki-01 committed:
♻️ update スマホ用の設定の変更

shiki-01 committed:
♻️ update 基本を iconSize に

shiki-01 committed:
♻️ update

shiki-01 committed:
📥 add タイトルとか説明の追加

shiki-01 committed:
🐛 fix 横並びの位置を変更できるように

shiki-01 committed:
📥 add 水平線忘れてた

wappon28dev reviewed (changes requested):
少しコード量が多いですね.  
追加したい機能は主に...

- `span.icon` を Iconify のアイコンに変更する
- コードブロックのハイライト
- 画像の属性の取得？
  な感じですね. microCMS からの HTML のパーサーを書いているようですが, 実は `lib/utils/middlewares` に rewriter というものがあります.

---

説明不足であれば申し訳ないのですが, rewriter は HTML を書き換える役割です.  
例えば, `add-anchor-copy.ts` と `add-anchor-copy.test.ts` を見てみましょう.

```ts
export function addAnchorCopy(html: string): string {
  const $ = load(html);
  const $header = $("h2, h3");
  // ...
```

`rewriteHTML` 関数から `addAnchorCopy` が呼ばれ, `html` に生の HTML が渡されます.  
`load` で HTML をパースし, `$` に格納されています. また, `$` は jQuery のように使えるので, `$("h2, h3")` で `h2` と `h3` を取得しています.
その後, HTML を書き換えて, `$.html()` で文字列に戻しています.

このように rewriter で HTML を書き換えてから Svelte でレンダーする流れを取ると, パーサーを書く必要がなく, もう少しコードが短くなると思います！

---

<details>
<summary>HTML からすべての画像の属性を得る</summary>

```ts
// `src/lib/utils/middlewares/rewriters/figure.ts` とか
import { load } from "cheerio";

export function processFigure(html: string): string {
  const $ = load(html);
  const $img = $("img");

  $img.each((_, elem) => {
    const $elem = $(elem);
    const img = $elem.attr("src");
    const alt = $elem.attr("alt");
    const width = $elem.attr("width");
    const height = $elem.attr("height");

    // ...

    $(elem).replaceWith(/* ... */);
  });

  return $.html();
}
```

</details>

---

<details>
<summary>また, Svelte コンポーネントを使いたいこともあるでしょう.</summary>

```ts
// `src/lib/utils/middlewares/rewriters/icons.ts` とか
// `span.icon` を Iconify のアイコンに変更する
import Icon from "@iconify/svelte";
import type { ComponentProps } from "svelte";
import logoTitle from "$lib/assets/icons/logo-title";

type IconProps = ComponentProps<Icon>["icon"];

const specialIcons: Record<string, IconProps> = {
  logoTitle: logoTitle,
};

export function injectIconify(): void {
  if (document == null) {
    throw new Error(
      "document is not defined. This fn should be called after `onMount`",
    );
  }

  document.querySelectorAll("span.icon").forEach((elem) => {
    const iconName = elem.textContent;
    if (iconName == null) {
      throw new Error("icon is not defined");
    }

    const icon: IconProps = specialIcons[iconName] ?? iconName;

    new Icon({
      target: elem,
      hydrate: true,
      props: { icon },
    });
  });
}
```

```ts
// `src/lib/utils/middlewares/rewrite-html.ts` とか適当な場所
export function injectComponents(): void {
  const arr: Array<() => void> = [injectIconify]; // 今後はここに注入したい Svelte コンポーネントを追加

  arr.forEach((fn) => {
    fn();
  });
}
```

の 2 つを用意して,

```svelte
// ...
onMount(() => {
  injectComponents();
});
```

とすればパーサーを書かなくても, Svelte コンポーネントを使うことができます！

</details>

---

な感じで...

1. パーサーを使わずに
1. ページ側の HTML の注入は次のように

   ```svelte
   <article>
     <div id="date-at">
       <p>公開日：{formatDate(data.detail.publishedAt)}</p>
       <p>更新日：{formatDate(data.detail.updatedAt)}</p>
     </div>
       <!-- eslint-disable-next-line svelte/no-at-html-tags -->
       {@html rewriteHTML(data.detail.content)}
   </article>
   ```

書き直してみると嬉しいなー！と.  分からないことがあればお聞きください！

shiki-01 committed:
Update +page.svelte

shiki-01 commented:
はっ…こんなものが…l
急いで書き換えたいと思いますっ

レビューありがとうございますです！

shiki-01 committed:
📁 save

shiki-01 commented on src/lib/utils/middlewares/rewrite-html.ts:
```diff
@@ -1,5 +1,15 @@
 import { highlight } from "./rewriters/highlight";
+import { injectIconify } from "./rewriters/icons";
+
+export function injectComponents(): void {
```
@wappon28dev 

書いてみたのですが、 Icon を追加しても、取得できないんですよね…

`src/lib/utils/middlewares/rewriters/icons.ts` 
```typescript
document.querySelectorAll("span.icon")    //null
document    //全部のHTML
```
となってしまい、表示ができず…どうすればいいでしょう？

wappon28dev replied:
ここらへんですね.

https://github.com/misc-org/misc-mdn.dev/blob/e20faf7adeae2261e16a57736a46029e2773d604/src/routes/blogs/%5Bslug%5D/%2Bpage.svelte#L17-L21

マウント時に,
1. (不要ですが) パースをする
2. コンポーネントの注入する
3. コンテンツをロードしたかのフラグ (`isContentLoaded`) を立てる
をしていますが,

https://github.com/misc-org/misc-mdn.dev/blob/e20faf7adeae2261e16a57736a46029e2773d604/src/routes/blogs/%5Bslug%5D/%2Bpage.svelte#L65-L67

を見ると, `isContentLoaded` が Falsy のとき, microCMS から渡された HTML は表示されないのではなく, DOM としてマウントされません.

つまり, コンポーネントの注入先がないってことです.
ちなみに, `isContentLoaded` は何に使っているのでしょうか？  不要であれば削除しちゃって問題ないかと思います！

wappon28dev replied:
ちなみに Prettier と ESLint が効いてないと思われます.  VSCode の拡張機能か, `yarn fmt` や `yarn lint` していただけると！

shiki-01 replied:
ありがとうございます！

ちなみになんですけど、
https://github.com/misc-org/misc-mdn.dev/blob/e20faf7adeae2261e16a57736a46029e2773d604/src/routes/blogs/%5Bslug%5D/%2Bpage.svelte#L65-L67
のところで、`{@html} can lead to XSS attack.` と出てしまうのですが、`dompurify` とかで一応サニタイズした方がいいですかね？

wappon28dev replied:
サニタイズの必要性ですが, 一般にインラインや script タグで js が含まれているなどが原因になります.
しかし, 今回その HTML の大本は microCMS で, 開発者が把握できる (他のユーザーの入力が一切含まれていない) ので, まあここはサニタイズしなくても良いかな～～～ という感じです.

https://github.com/misc-org/misc-mdn.dev/blob/1b44558e3372882e74be91a9d9db43c39d7ad5a3/src/routes/blogs/%5Bslug%5D/%2Bpage.svelte#L39-L40

(また, 今回 SSG するので, ビルド後は `@html` の箇所は注入されたあとの状態になりますし)
心配なのであれば, CSP など設定しておくと良いかも と思いました.

shiki-01 committed:
🐛 fix fmt と その他もろもろ

- service/hifhlight.ts の削除
- icon.ts でサイズ変更をできるように
- blogs/[slug]/+page.svelte で表示ができるよう不要な部分を削除

shiki-01 committed:
Update icons.ts

shiki-01 committed:
🗑️ remove 不要物の削除

wappon28dev commented on docs/microcms/tips.md:
````diff
@@ -0,0 +1,124 @@
+
+### 新たな追加方法
+
+新たなHTML要素をパースする機能を追加するには、以下の手順を守ってください。
+
+1. **型定義の追加：** まず、新たなHTML要素を表現するための型を定義します。これは、既存の型定義を参考にして行ってください。
+2. **パーサー関数の作成：** 次に、新たなHTML要素を解析するための関数を作成します。この関数は、HTML要素を受け取り、1で定義した型のオブジェクトを返すようにします。
+3. **メインパーサーの更新：** 最後に、メインのパーサー関数を更新して、新たに作成したパーサー関数を呼び出すようにします。これにより、新たなHTML要素が適切に解析され、特定の形式に変換されます。
+4. **表示の更新：** 最後に、[`src/routes/blogs/[slug]/+page.svelte` ↗](src/routes/blogs/[slug]/+page.svelte) や [`src/lib/components/BlogParser.svelte` ↗](src/lib/components/BlogParser.svelte) などを更新して、パースしたものが正しく描写されるようにしてください。
+
+カスタムクラスを追加するときは、基本的に `<span class="custom-class"></span>` が `p` タグの中に入っているので、`className` によって条件を分岐させてください。
+
+具体的には、下記の部分を変更します。
+
+[`parser.ts` ↗](src/lib/utils/services/parser.ts)
+
+```diff
+- type SpanElement = { type: 'icon', content: string, size?: number }
+
++ type SpanElement =
++   | { type: 'icon', content: string, size?: number }
++   | { type: 'newClass', content: Content };
+```
+
+また、`processSpan` 関数で下記の部分に追記し、条件の分岐を行います。
+
+```diff
+   if (className === 'icon') {
+          const match = content.match(/\$\$(\d+)$/);
+           if (match) {
+              content = content.replace(/\$\$\d+$/, '');
+              size = parseInt(match[1], 10);
+        }
+          return { type: 'icon', content, size };
+-  }
+
++  } else if (className === 'newClass') {
++      //新しい処理
++  }
+```
````
パーサーについての補足は不要かもです！

wappon28dev commented on docs/microcms/tips.md:
```diff
@@ -0,0 +1,124 @@
+
+microCMS のカスタムクラスに、アイコンというものが追加されているので、 [icones ↗](https://icones.js.org) から使いたいアイコンの `mid:` などの部分をコピペして、張り付けた部分をアイコンというクラスに設定してください。
+
+基本的にはサイズは文字の大きさに合わせていますが、文字列の最後に `$$00` のように `$$` ＋ 半角数字で指定することで大きさを変えることができます。
+なお、**この大きさは相対値**なので「基本のサイズプラスいくつ」で考えて作成してください。
+
+ex)
+
+通常
+
+- `mdi:github` → microCMS でクラスを「アイコン」に設定
+
+文字サイズを指定する場合
+
+- `mdi:github$$40` → microCMS でクラスを「アイコン」に設定 ＋ サイズを縦横 40 に設定
+
```
ここも……？

挙動をこのようにするならば, このままでも良いかもです.

shiki-01 replied:
一応

https://github.com/misc-org/misc-mdn.dev/blob/3a7d0b7535530ead0722d3065c1ac48c817f3600/src/lib/utils/middlewares/rewriters/icons.ts#L26-L31

の部分でアイコンサイズの変更はできるようにしてあるんので、こちらはこのままにしようかなと！

（記法だけ少し変わったので修正しました）

shiki-01 committed:
Update tips.md

wappon28dev commented on package.json:
```diff
@@ -18,6 +18,8 @@
   "devDependencies": {
     "@iconify/svelte": "^3.1.6",
     "@markuplint/svelte-parser": "^4.4.0",
+    "@skeletonlabs/skeleton": "^2.9.2",
+    "@skeletonlabs/tw-plugin": "^0.4.0",
```
この 2 つの依存関係はもはや使われていないようです.
- `$ npm remove @skeletonlabs/skeleton`
- `$ npm remove @skeletonlabs/tw-plugin`
で削除しちゃいましょう.

wappon28dev commented on docs/microcms/tips.md:
```diff
@@ -0,0 +1,80 @@
+| 埋め込み       | 外部コンテンツを埋め込みます。                                 |
+| アイコン       | [icones ↗](https://icones.js.org)にあるアイコンを表示します。 |
+
+### テーブルについて
+
+テーブルに関しては、処理の複雑さや容量の増加を考えて、`TextFormat` （ [`parser.ts` ↗](src/lib/utils/services/parser.ts) の型宣言の部分にあります ） にある基本的なものと画像のみとなっています。
+
```
```suggestion
```

`parser.ts` はもう無いので, ここの記述は削除していただけると！

wappon28dev commented on src/routes/blogs/[slug]/+page.server.ts:
```diff
@@ -1,11 +1,29 @@
+    return {
+      detail: res,
+    };
+  } catch (error: any) {
+    if (error.response?.status === 404) {
+      return {
+        status: 404,
+        error: new Error("Not Found"),
+      };
+    }
+  }
```
この記述だと, 403 や 500 などのエラーを握りつぶしていまいます. if 文のあと, もう一度 `error` を Throw すると良いかもです.
また, try-catch 文をなくした場合でも `+error.svelte` がエラーページを描画してくれる (はず？) ので, そもそも try-catch 文は不要かもです.

wappon28dev commented on src/routes/blogs/[slug]/+page.server.ts:
```diff
@@ -1,11 +1,29 @@
+    if (draftKey) {
+      res = await getDraftContentDetail("blogs", slug, draftKey);
+    } else {
+      res = await getContentDetail("blogs", slug);
```
もはや仕様変更レベルを見逃していてとってもスミマセンですが, 1 つ良いでしょうか？

現在, microCMS へのアクセスを減らすためにも, 高速化のためにも, 現在 `/blogs/[slug]` を含めたすべてのページが Pre-render されます.  
https://github.com/misc-org/misc-mdn.dev/blob/3a7d0b7535530ead0722d3065c1ac48c817f3600/src/routes/%2Blayout.ts#L1

しかし, Pre-render されてほしいページは既に公開されているページ (microCMS で “公開” したページ) です.  
そのことに対して画面プレビューもこのページ (厳密にはパス) で処理させる場合, その仕組み上 microCMS へのアクセスが必要で, Pre-render の意図と矛盾してしまいます.  
ちなみに現在の状態だと, Svelte の Pre-render**er** がこのページを走査してビルドするとき, `draftKey` パラメーターは付与せずレンダーするので, microCMS から “画面プレビュー” しても, どのページでも Not Found を返します！

まとめますと,
- `/blogs/[slug]` は Pre-render するようにして, `getContentDetail` を返す
- `/drafts/blogs/[slug]?draftKey=...` は Pre-render しないようにして, `getDraftContentDetail` を返す
のような設計にすると良いと思います！

パスによって Pre-render するかしないかの設定は [Page options - SvelteKit Docs ↗](https://kit.svelte.dev/docs/page-options#prerender) を参考にすると良いと思います！

shiki-01 replied:
一応自分なりに書いてみましたがもし何かあれば…！

see:
- [`src/routes/+layout.ts`](https://github.com/misc-org/misc-mdn.dev/blob/add/highlight/src/routes/%2Blayout.ts)

wappon28dev commented on docs/microcms/tips.md:
```diff
@@ -0,0 +1,80 @@
+
+microCMS では、画面プレビューという機能を用いて実際に投稿したときのプレビューを見ることができます。投稿の前に必ず確認して、正しく書けているか・許容されているフォーマットかどうかなどを確認してください。
+
+---
+
+## パーサーについて
+
+この章は、[`parsr.ts` ↗](src/lib/utils/services/parser.ts) を参照しながら読んでください。
+
+### 型定義
+
+まず、HTML要素を表現するための型が定義されています。これらは、HTMLの各要素（段落、見出し、リスト、テーブルなど）を表現するためのものです。例えば、`MainElement` 型は、HTMLの主要な要素を表現します。これらの型は、パーサーがHTMLを解析し、特定の形式に変換する際に使用されます。
```
```suggestion

```

ここの記述も古くなっているかもです.

wappon28dev reviewed (changes requested):
マージまであとちょっとかと思います！  もう少しお願いします！

shiki-01 committed:
🗑️ remove 不要パッケージ削除とテーブルに関する記述の削除

shiki-01 committed:
Update +page.server.ts

shiki-01 committed:
🐛 fix draftKeyがあるときは prerender を false に

shiki-01 committed:
🐛 fix run fmt

shiki-01 committed:
🐛 fix npm run lint + @htmlの 無視

shiki-01 committed:
Update tips.md

wappon28dev commented on src/routes/+layout.ts:
```diff
@@ -1 +1,5 @@
-export const prerender = true;
+export const prerender = (request: { url: string | URL }) => {
+  const url = new URL(request.url);
+  const draftKey = url.searchParams.get("draftKey");
+  return !draftKey;
+};
```
また遅くなってしまいました.  すみません. 🙇 

---

おっと.  あまり見慣れない書き方をしているようですね！  
この `const prerender = ...` というものはあくまで SvelteKit に Pre-render するかしないかの設定値を公開しているだけなので, 関数が呼び出されたりすることも, `request` に URL が入ったりすることはありません.  ためしに, `$ npm run build` などしてみてください.  おそらくビルドがコケると思います.

前回のコメントどおり,

> まとめますと,
> - `/blogs/[slug]` は Pre-render するようにして, `getContentDetail` を返す
> - `/drafts/blogs/[slug]?draftKey=...` は Pre-render しないようにして, `getDraftContentDetail` を返す
> 
> のような設計にすると良いと思います！

を実装してみましょう.  
具体的には, `/routes/blogs/[slug]/+page.svelte` や `/routes/drafts/blogs/[slug]/+page.svelte` を作る感じです！  分からなくなってしまったらぜひお伝えください！  私が代わりにコミットしようと思います.

see: [Page options - SvelteKit Docs ↗︎](https://kit.svelte.dev/docs/page-options#prerender)

shiki-01 replied:
レビューありがとうございます！
いろいろやってみたのですが、そもそも `/routes/blogs/[slug]/+page.svelte` が Pre-render してくれなく…

もしよかったらコミットお願いしてもいいでしょうか X(

shiki-01 committed:
draft/blogs/[slug]/+page.svelteの追加

shiki-01 committed:
不要部分の削除

wappon28dev committed:
🔥 ページネーションの削除とリファクタリング

wappon28dev committed:
✨ 下書きモード

wappon28dev committed:
🔥 `prerender.handleHttpError`

wappon28dev reviewed (commented):
こんにちは.  数日お待たせしてすみませんでした. 🙇 

パス `/routes/blogs/[slug]/+page.svelte` が Pre-render できない理由を調査をしたところ, Blog 一覧を表示する BlogPage コンポーネントが原因だと分かりました.

ページネーションのための Limit が悪さをしているようです:

https://github.com/misc-org/misc-mdn.dev/blob/72fdc38b7af038c08167786ebcdbeea29eff6022/src/lib/components/BlogPage.svelte#L23-L24

このコードは, `page` に初期値である `1` が代入されているので, このコンポーネントの `onMount` が呼び出されるまで `0` になってしうことで, For が回らなく, 記事が 1 つも無い判定になっていたようです. (実際, ページロードのときに `記事がありません` が見えますね.)

これがどう Pre-render に響いていくのかというと, ズバリ Pre-render の仕組みが関わっています. 公式ドキュメント ([Page options - SvelteKit Docs ↗](https://kit.svelte.dev/docs/page-options#entries)) によると,

> ...SvelteKit will prerender `/` and `/blog`, and in the process discover links like `<a href="/blog/hello-world">` which give it new pages to prerender.

のように記載があり, `<a>` タグの発見可能が必要 → `BlogLink` コンポーネントの描画が必要 → `BlogPage` コンポーネントで `BlogLink` コンポーネントの For が必要 → `limit` の初期値が `0` でない... のような感じに繋がっていることが分かりました.

これを解決するために, ページネーションを削除し, 初期状態ですべての記事がちゃんと描画されるようにするコミットを積んでおきました！ 🙌 
あと, コンポーネント名や記法の変更を主とするリファクタリングもしておきました！ ♻️ 

あとは, Draft 機能が不完全であったため, これも Pre-render されるように記事表示のコンポーネントを `/blogs/[slug]` と共通化し, 適切なエラーハンドリングも追加しておきました！

@shiki-01 が積んだコミットの最後から, 私の積んだコミットの最後までの File Changes を見て, 不明瞭や不明なコードがあれば, ぜひお聞きくださいね.

shiki-01 committed:
✨ create Dummy Image の追加

shiki-01 commented:
確認しました！
自分が色々謎の設定をしてしまっていたようですね…勉強になります。
そして、お忙しいなかコミットもしていただきありがとうございます！

全体的に理解できたので、私からは特にないですが、既知の問題だったダミー画像を作成しておきました。
コミット確認いただけると！

wappon28dev committed:
💫 DummyImage の Hover を設定

wappon28dev reviewed (approved):
DummyImage のコミット OK です！  LGTM!
ついでに, Hover 時のアニメーションを普通の画像とおなじように設定しておきました！

長い間お疲れ様でした！  仕様の練りが足りなかったのか, 何度かコメントしたり, Reject したりして申し訳なかったです.

近いうちに, GitHub Actions を使って本番環境へデプロイしたいですね.  Figma でのデザインから着実に実装してくれたので, 立派なサイトになりそうです.
